brainstorming.md

```diff
@@ -0,0 +1,372 @@
+>
+> **Proposal**: restrict `uns` to a mapping of string keys and `simple` values. Extend to complex types in the future as we have concrete use cases which warrant the complexity.
+
+> ⚠️ **Issue** - the name `uns` is not really all that suggestive. **Proposal:** rename `uns` to `metadata`, eg, `print(soma.metadata['version'])`
+
+## SOMACollection
+
+SOMACollection is an unordered, `string`-keyed map of values of type SOMADataFrame, SOMAMatrix, SOMACollection or SOMA. SOMACollection objects are typically persistent, and may be nested to any level (eg, SOMACollection may contain other SOMACollection objects). Keys in the map are unique and singular (no duplicates, ie, it is not a multi-map).
+
+## SOMADataFrame
+
+SOMADataFrame is a 1-D multi-column table of monomorphic arrays, all of equal length -- ie, a "dataframe". The SOMADataFrame has a user-defined schema, which includes:
+
+- a table schema, expressed as an `Arrow.RecordBatch` type, defining number of columns, and their respective names and types
+- the index column name
+
+A SOMADataFrame must contain at least a single column which is the named index column. The index column _must_ be of `string` type. Index values are non-null and must be unique (ie, no duplicates are allowed) within the object. The index is sparse and unordered, ie, supported indexing is by ID value only, with no indexing order or positional (offset-based) indexing.
+
+Most language-specific implementations will present SOMADataFrame in more convenient forms, such as Python `pandas.DataFrame`, R `data.frame` or other environment-specific data structures. Specific implementations of the SOMA API may extend this type with convertors to/from the underlying data structure, and enforce indexing conventions.
+
+> ⚠️ **Issue** - this spec requires that index column is a `string` type. Most underlying "engines" will support more flexibility. Do we want to enable that flexibility, or stick with requiring string for simplicity/commonality?
+
+## SOMAMatrix
+
+SOMAMatrix is a 2-D multi-column table of monomorphic arrays, ie, a "dataframe" with two index dimensions. SOMAMatrix is an extension of SOMADataFrame, with two indexing dimensions. The SOMAMatrix has a user-defined schema, which includes:
+
+- a table schema, expressed as an `Arrow.RecordBatch` type, defining the number of columns, and their respective names and types
+- the _primary_ index column name
+- the _secondary_ index column name
+
+A SOMAMatrix must contain (minimally) two columns which are the primary and secondary index columns. The index columns _must_ be of `string` type. Index values are non-null and must be unique (ie, no duplicates are allowed) within the object. The index is sparse and unordered, ie, supported indexing is by ID value only, with no indexing order or positional (offset-based) indexing.
```

Comment:
the question of how many attribute columns is omitted here -- maybe best --

i would go ahead & add that _nominally_ there will be a single attribute -- this will help the reader 'aha, i know what this is for'

Reply:
Good catch - we need to add clarity here.

BTW, in this API, there is no concept of an "attribute column" (I assume you are using that in the TileDB sense of the term).  Rather, there are N>=3 columns, and two of them are designated the index columns.  Underlying implementations need to figure out how to map that to their schema (which in the TileDB engine would simply mean the two named index columns are dims, the rest are attributes).

Reply:
Fix added to both SOMADataFrame and SOMAMatrix.